### PR TITLE
ci: add problem matchers

### LINKED ISSUE #2661: problem matcher not working on build matrix
### Describe the bug

BN: https://github.com/cataclysmbnteam/Cataclysm-BN/actions/runs/4722885326/jobs/8378085899?pr=2570

DDA: https://github.com/CleverRaven/Cataclysm-DDA/actions/runs/4721579034/jobs/8375027698?pr=64725
image
https://github.com/CleverRaven/Cataclysm-DDA/pull/64725/files

### Steps To Reproduce
1. fail a test
2. it does not show up on problem matcher

### PR DESCRIPTION
**Pull request by scarf005**
#### Summary

SUMMARY: Infrastructure "Port DDA ci problem matchers"

#### Purpose of change
- fix #2661 

#### Describe the solution

ported
- https://github.com/CleverRaven/Cataclysm-DDA/commit/493f6d4e83ffa2ca62133f96dc3a9b8851a5bc4a by @Qrox 
- https://github.com/CleverRaven/Cataclysm-DDA/pull/44111 by @jbytheway 
- https://github.com/CleverRaven/Cataclysm-DDA/pull/44126 by @jbytheway 

added
- https://github.com/marketplace/actions/msvc-problem-matcher

[flake8](https://github.com/CleverRaven/Cataclysm-DDA/pull/44192) is omitted due to BN migrating to typescript for tool scripts.

#### Describe alternatives you've considered

cry in pain

#### Additional context

TODO: add android problem matchers: https://github.com/marketplace/actions/android-problem-matchers

### CODE REVIEW

**Comment by Qrox** [comment]
IIRC this fix is for displaying multi-line JSON errors, but the issue you linked is about unit test (c++) errors.

**Comment by Qrox** [comment]
https://github.com/CleverRaven/Cataclysm-DDA/tree/master/build-scripts/problem-matchers is probably what you are looking for because I don't see the files for problem matchers in the BN repo. The relevant PRs are CleverRaven/Cataclysm-DDA#44111, CleverRaven/Cataclysm-DDA#44126, and CleverRaven/Cataclysm-DDA#44192. Problem matchers only support single-line error messages though, at least when I last checked.

File: src/json.cpp [hunk 1]
```diff
@@ -1701,16 +1702,112 @@ bool JsonIn::read( JsonDeserializer &j, bool throw_on_error )
...
+ * See https://github.com/actions/toolkit/blob/main/packages/core/src/command.ts
+ */
+static std::string escape_data( std::string str )
+{
+    switch( error_log_format ) {
+        case error_log_format_t::human_readable:
+            break;
+        case error_log_format_t::github_action:
+            replace_all( str, "%", "%25" );
+            replace_all( str, "\r", "%0D" );
```
Comment (Coolthulhu) [inline src/json.cpp:1783]: You may want to silence the `cata-text-style` warning here.

File: src/json.cpp [hunk 2]
```diff
@@ -1701,16 +1702,112 @@ bool JsonIn::read( JsonDeserializer &j, bool throw_on_error )
...
+ * See https://github.com/actions/toolkit/blob/main/packages/core/src/command.ts
+ */
+static std::string escape_property( std::string str )
+{
+    switch( error_log_format ) {
+        case error_log_format_t::human_readable:
+            break;
+        case error_log_format_t::github_action:
+            replace_all( str, "%", "%25" );
+            replace_all( str, "\r", "%0D" );
```
Comment (Coolthulhu) [inline src/json.cpp:1763]: You may want to silence `cata-test-style` here.

**Pull request merged**